### DISCUSSION TITLE
connecting acebase clients locally on server, with https

### DISCUSSION THREAD

**Discussion by dtau00** [appy-one/acebase-client#63] [Q&A]
this might be a dumb question, but its not making sense as to why its not working.  So i have a db set up on a server (localhost, https).  I have other apps running on the server that im trying to connect locally.  Throws a simple connection error.  Ive also tried various combinations of replacing localhost with 0.0.0.0.  Notes on testing:
  - running on a windows machine, private firewall is disabled
  - with https enabled, i can still connect to the webmanager.  but the client doesn't connect to the local server db.
  - disabling https works
  - connecting the client db to the domain name works (using cloudflare, although i think data is unnecessarily routed to the cloudflare server then back to the server itself)

i guess my first question is, is this suppose to work?  if so, do you think its a networking or https issue?  i can keep searching for a solution with some direction.  thanks!

**Comment by appy-one** [comment]
Can you maybe post some code?

**Comment by dtau00** [comment]
// initializing acebase server
const dbname = 'trading_db'
const options = {  
    host: '0.0.0.0', port: 2053, // also tried with localhost. but i need to use 0.0.0.0 for it to work.
    https: {
      certPath: '../certificates/client-cert.pem',
      keyPath:  '../certificates/client-key.pem',
    },
    authentication: {
        enabled: true,
        allowUserSignup: true,
        defaultAccessRule: 'auth'
    },
    transactionLogging: true, 
    transactions: { log: true, maxAge: 30, noWait: false },
    logLevel: 'warn', 
    autoGrow: true,
    storage: { path: '.', removeVoidProperties: true } ,
    email: { send: sendRequestedEmail }
};
const server = new AceBaseServer(dbname,options);

//iniialize acebase client on node apps
const host = config.get('db.host') // tried: localhost, 0.0.0.0, db.domain.com. only domain name works.
const port = config.get('db.port') // 2053
const dbname = config.get('db.dbname') // trading_db
const https = config.get('db.https') // true
export const username = config.get('db.username') // admin
export const password = config.get('db.password')
export const db = new AceBaseClient({ host: host, port: port, dbname: dbname, https: https, logLevel: 'verbose'});


log on client
[34m[trading_db][39m Connecting to AceBase server "https://localhost:2053"
[34m[trading_db][39m Using websocket transport for socket.io
[34m[trading_db][39m Websocket connection error: Error: websocket error


if i simply switch to point the client app (running on the local server) to the domain name, everything works.  if i turn off https then the localhost will work as expected.

**Comment by AAAATTIEH** [comment]
@appy-one is there a fix for this problem?

**Comment by appy-one** [comment]
Apologies for the late reply!

Why would you want to use SSL on localhost? SSL certificates are bound to domain names, not IP addresses. Since localhost is an alias for the loopback address 127.0.0.1, there is no need to an SSL certificate either because all communication takes place on a single machine.

If you allowing remote access to your server from a domain name, create an SSL certificate for it. When running locally only, just use http!

**Reply by dtau00** [reply]
what if the db has to be accessed publicly and internally?  for example, the client app is connecting to the db via https, but there are services on the local server that also accesses the db?

**Reply by AAAATTIEH** [reply]
Right now I couldn't connect locally on 127.0.0.1 in production mode so I have to connect via https with SSL certificate from the local server to the database which is hosted on the same local server. This should have a fix as fixing it will result in better performance on time of data retrieval/updating on local server.

**Reply by dtau00** [reply]
have you tried connecting to 0.0.0.0 via http?  that worked for me.

**Reply by AAAATTIEH** [reply]
The thing is I want to access the database from outside via https and at the same time want to access it locally. I've tried so many ways one of them was to create 2 servers one locally and one for remote connection via https but this created so many problems. I tried also connecting to localhost with https but this doesn't work also. So lastly I did what I have said earlier which is connecting locally via https with the domain name I have registered (not localhost) but this method has a drawback which is fetching/updating data time.

**Reply by dtau00** [reply]
oh, yea its the same problem i have.  thought you were just trying to connect internally.
